integreat_cms/api/v3/chat/utils/chat_bot.py

```diff
@@ -87,44 +91,50 @@ async def async_process_user_message(
 
 @shared_task
 def process_user_message(
+    message_timestamp: datetime,
     region_slug: str,
     zammad_ticket_id: int,
 ) -> None:
     """
     Call the async processing of the message from an Integreat App user. The question
     should be translated into the main language and an automatic answer should be generated.
+
+    :param message_timestamp: Timestamp of the article that triggered the webhook
+    :param region_slug: region used in webhook
+    :param zammad_ticket_id: Ticket ID contained in webhook request
     """
     region = Region.objects.get(slug=region_slug)
     zammad_chat = UserChat.objects.get(zammad_id=zammad_ticket_id, region=region)
-    client = ZammadChatAPI(region)
-    messages = client.get_api_response(zammad_chat)["messages"]
-    if not isinstance(messages, list):
-        raise TypeError("Messages is are not a list")
     translation, answer = asyncio.run(
         async_process_user_message(
             zammad_chat.language.slug,
             region_slug,
             region.default_language.slug,
-            messages,
+            # Prevent a race condition where new articles can be added to a ticket
+            # while the webhook has not yet sent to and processed by the Integreat CMS
+            [
+                message
+                for message in zammad_chat.messages
+                if datetime.fromisoformat(message["created_at"]) <= message_timestamp
+            ],
         ),
     )
     if translation:
-        client.send_message(
-            zammad_chat.zammad_id,
-            translation["translation"],
-            True,
-            True,
+        zammad_chat.save_message(
+            message=translation["translation"],
+            internal=True,
```

Comment:
So from what I gather, internal mostly means "user should not be able to see the message", correct? Unless there's also a secondary purpose, I think a more descriptive name might be helpful

Reply:
Yes, internal indicates that the message is only visible to counselors (Zammad agents). The name is derived from the corresponding Zammad article attribute.